File: .github/scripts/sqlite-online-backup.py
#!/usr/bin/env python3
"""SQLite Online Backup API via Python's sqlite3.Connection.backup()."""
import json
import sqlite3
import sys


def integrity_check(db_path: str) -> str:
    conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
    try:
        return str(conn.execute("PRAGMA integrity_check").fetchone()[0])
    finally:
        conn.close()


def main() -> int:
    if len(sys.argv) == 3 and sys.argv[1] == "integrity":
        print(json.dumps({"integrity_check": integrity_check(sys.argv[2])}))
        return 0
    if len(sys.argv) != 3:
        print("usage: sqlite-online-backup.py <src.db> <dest.db>", file=sys.stderr)
        print("   or: sqlite-online-backup.py integrity <db>", file=sys.stderr)
        return 2
    src, dest = sys.argv[1], sys.argv[2]
    source = sqlite3.connect(f"file:{src}?mode=ro", uri=True)
    try:
        dest_conn = sqlite3.connect(dest)
        try:
            source.backup(dest_conn)
        finally:
            dest_conn.close()
    finally:
        source.close()
    print(json.dumps({"ok": True, "method": "python3 sqlite3.Connection.backup"}))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .github/scripts/sqlite-online-backup.mjs
#!/usr/bin/env node
/**
 * SQLite Online Backup API via node:sqlite when the runtime exports backup().
 * Node 22.14 (current CI / likely Production image) does not export backup().
 * Callers must capability-detect and fall back to python3 Connection.backup
 * or sqlite3 CLI `.backup` — never raw-copy a live DB.
 */
import * as sqlite from "node:sqlite";

export function backupApiAvailable() {
  return typeof sqlite.backup === "function";
}

export async function onlineBackup(srcPath, destPath) {
  if (!backupApiAvailable()) {
    const err = new Error("node:sqlite backup() is not available");
    err.code = "BACKUP_API_UNAVAILABLE";
    throw err;
  }
  const sourceDb = new sqlite.DatabaseSync(srcPath, { readOnly: true });
  try {
    const pages = await sqlite.backup(sourceDb, destPath);
    return { ok: true, pages, method: "node:sqlite backup()" };
  } finally {
    sourceDb.close();
  }
}

const self = process.argv[1] ? String(process.argv[1]) : "";
if (self.endsWith("sqlite-online-backup.mjs")) {
  const src = process.argv[2];
  const dest = process.argv[3];
  if (!src || !dest) {
    console.error("usage: sqlite-online-backup.mjs <src.db> <dest.db>");
    process.exit(2);
  }
  if (!backupApiAvailable()) {
    console.error("BACKUP_API_UNAVAILABLE");
    process.exit(3);
  }
  const result = await onlineBackup(src, dest);
  console.log(JSON.stringify(result));
}
